### PR TITLE
Add methods for RouteConfig ITN

### PR DESCRIPTION
**Pull request by ecfm**
http://builds.cask.co/browse/CDAP-DUT4953-11

### CODE REVIEW

File: cdap-client-tests/src/test/java/co/cask/cdap/client/AbstractClientTest.java
Comment (bdmogal) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/AbstractClientTest.java]: deprecation in test classes? Are these APIs? Can you just remove this in favor of the other method?

**Review by bdmogal — COMMENTED** [review]
@maochf why do we need a separate RouteConfigClient? It does not seem like it should have its own client class. It can only be used with Services right? Why not add methods to ServiceClient instead?

**Comment by ecfm** [comment]
I think we need a `RouteConfigClient` because we have a separate `RouteConfigHttpHandler` and `ServiceClient` is used for a single version of a service but `RouteConfigClient` is for all the versions of a service.

**Comment by bdmogal** [comment]
but shouldn't ServiceClient be adopted to support versions instead?

**Comment by bdmogal** [comment]
Oh, but you need these client APIs to add/update routes? I see, then it may be ok.

**Comment by ecfm** [comment]
Build is green http://builds.cask.co/browse/CDAP-DUT4953-5/test

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java:515]: are these new APIs reviewed/documented? Also, will the old APIs always continue to exist?

File: cdap-client-tests/src/test/java/co/cask/cdap/client/AbstractClientTest.java
Comment (bdmogal) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/AbstractClientTest.java]: instead of accepting a String, accept `ProgramRunStatus`?

File: cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java
Comment (bdmogal) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java]: nit: remove extra newline

File: cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java
Comment (bdmogal) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java]: `NamespaceId.DEFAULT.artifact("appreturnsArgs", "1.0.0");`

File: cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java
Comment (bdmogal) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java]: `new HashMap<>()`

File: cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java
Comment (bdmogal) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java]: this method's contract is strange. Can you have a method that deploys the app and a different method to start the service? 
The method that starts a service should accept a map of runtime arguments. As a result, you don't need to return the runtime arguments from that method.

File: cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java
Comment (bdmogal) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java]: the `<String, Integer>` is not needed in any of these method calls

File: cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java
Comment (bdmogal) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java]: remove this single-line method?

File: cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java
Comment (bdmogal) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java]: ditto

File: cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java
Comment (bdmogal) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java]: if the route config is invalid, you should get a better exception, right? Like a BadRequestException

File: cdap-client/src/main/java/co/cask/cdap/client/RouteConfigClient.java
Comment (bdmogal) [inline cdap-client/src/main/java/co/cask/cdap/client/RouteConfigClient.java]: use single space after parameter name. applies to all javadocs in this class

File: cdap-client/src/main/java/co/cask/cdap/client/RouteConfigClient.java
Comment (bdmogal) [inline cdap-client/src/main/java/co/cask/cdap/client/RouteConfigClient.java]: call this `buildRouteConfigUrl`?

File: cdap-integration-test/src/main/java/co/cask/cdap/test/remote/RemoteRouteConfigManager.java
Comment (bdmogal) [inline cdap-integration-test/src/main/java/co/cask/cdap/test/remote/RemoteRouteConfigManager.java]: why wrap everything as a runtime exception? what exceptions are thrown here?

**Review by bdmogal — CHANGES_REQUESTED** [review]
@maochf took one pass. However, I still believe that there shouldn't be separate client and test manager classes for route config. Route config is extremely tightly coupled with Services - since we have those classes for Services, we should just add to them. But feel free to get another opinion too.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java
Comment (ecfm) [reply cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java:515]: Old API's need to exist for backward compatibility and users not using versions. It is not documented or reviewed yet. I can remove it from this PR since it is not actually used.

Comment (bdmogal) [reply cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java:515]: didn't mean for it to be removed. It seems like a useful API, but my question is what is the plan for the old APIs? Will they continue to exist for ever or are they just for backward compatibility. If it is just for backward compatibility, then those APIs should be deprecated.

Comment (ecfm) [reply cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java:515]: They will probably be there forever. I assume that not every user need versions for their apps. Need to confirm this with Sree.

File: cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java
Comment (ecfm) [reply cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java]: propMap is returned because it shares the same content except for the app version. It is constructed within this method to avoid duplicated code.

File: cdap-integration-test/src/main/java/co/cask/cdap/test/remote/RemoteRouteConfigManager.java
Comment (ecfm) [reply cdap-integration-test/src/main/java/co/cask/cdap/test/remote/RemoteRouteConfigManager.java]: Trying to follow the style of `RemoteApplicationManager#getHistory` Should I unwrap it and let it throw exceptions directly?

File: cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java
Comment (ecfm) [reply cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java]: `RouteConfigHttpHandler` throws BadRequestException but `RouteConfigClient` can only get IOException from the HTTP response. BadRequestException can be read from the HTTP status code from the response message.

Comment (bdmogal) [reply cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java]: RouteConfigClient should throw BadRequestException if it gets a 400. The message in the exception should be the response body.

File: cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java
Comment (bdmogal) [reply cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java]: but to avoid duplicate code, this method becomes weird: Why should something that starts a service return the runtime arguments? It seems like runtime arguments should be provided by the caller of the method that starts the service.

**Comment by ecfm** [comment]
I just realized that `RouteConfigManager` should not be combined with `ServiceManager` because after combining, `DefaultServiceManager` used in unit tests also need to implement RouteConfig related methods, but we decided that RouteConfig tests should only be done in integration tests.

File: cdap-integration-test/src/main/java/co/cask/cdap/test/IntegrationTestBase.java
Comment (bdmogal) [inline cdap-integration-test/src/main/java/co/cask/cdap/test/IntegrationTestBase.java]: so this is wrong, right? 

The method says it will return RouteConfigManager, but it always returns RemoteRouteConfigManager. So technically, you can even call `getRouteConfigManager` in unit tests, but it will return the remote implementation? Shouldn't we let the `TestManager` class decide which implementation to use based on the runtime?

File: cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java
Comment (bdmogal) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java]: nit: remove this line

File: cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java
Comment (bdmogal) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java]: is there a `ServiceClientTestRun` that you can move these tests to?

File: cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java
Comment (bdmogal) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java]: just `startService`. 

Also, this method still returns runtimeArgs? I thought we discussed that it should take the service and the runtime args, and then start it?

**Review by bdmogal — CHANGES_REQUESTED** [review]
Couple minor comments.

File: cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java
Comment (ecfm) [reply cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java]: right, forgot to pop the change from stash

File: cdap-client/src/main/java/co/cask/cdap/client/ProgramClient.java
Comment (bdmogal) [inline cdap-client/src/main/java/co/cask/cdap/client/ProgramClient.java]: shouldn't change this, since it is incompatible. Instead, this should do `waitForStatus(program.toEntityId(), ProgramStatus.valueOf(status), timeout, timeoutUnit);`

File: cdap-client/src/main/java/co/cask/cdap/client/ProgramClient.java
Comment (bdmogal) [inline cdap-client/src/main/java/co/cask/cdap/client/ProgramClient.java]: no need to store this, right? can just use `status.name()`

File: cdap-client-tests/src/test/java/co/cask/cdap/client/AbstractClientTest.java
Comment (bdmogal) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/AbstractClientTest.java]: use `name()` instead of `toString()`

**Review by bdmogal — CHANGES_REQUESTED** [review]
Just a couple more minor comments.

**Comment by bdmogal** [comment]
LGTM, as long as build succeeds. Please squash commits into one before merging.

**Review by bdmogal — APPROVED** [review]

**Comment by ecfm** [comment]
Build is green

**Pull request merged**
